### ISSUE TITLE
Node config issue with nvm and VS Code on Linux or macOS

### ISSUE THREAD

**Issue by sukheja-varun** [cyrilletuzi/vscode-angular-schematics#51]
Unable to execute any command using the plugin, getting below error

```
ng g enum abc --lintFix
/usr/bin/env: ‘node’: No such file or directory
```

I am using nvm (node version manager) instead of default node installed on root. All node versions are present inside .nvm folder.
So might be this plugin is not getting the node path.

So how do I define my node path for this plugin ??

**Comment by cyrilletuzi** [comment]
Thanks for your feedback.

The extension is not using your global Node. For its own code, the extension being inside VS Code, it uses the Node version included in VS Code (as it's an Electron app).

I suppose it's the Angular CLI which can't find your Node installation. Are the `ng g` commands working when you launch them directly inside the VS Code terminal?

**Comment by sukheja-varun** [comment]
Yes all ng commands works in vs code terminal, as I have set the terminal as zsh in my vs code settings and zsh loads default node mentioned by nvm.

**Comment by cyrilletuzi** [comment]
The extension is already taking the VS Code terminal config into consideration, so it's strange.

Do you confirm:
- you're using macOS or Linux?
- that `terminal.integrated.shell.osx` or `terminal.integrated.shell.linux` are set in your VS Code config?
- that this config is applied to the workspace where you're trying to use the extension?

**Comment by sukheja-varun** [comment]
Below are the details:-
- Ubuntu 18.10
- "terminal.integrated.shell.linux": "/usr/bin/zsh",
- I am able to use zsh terminal in my workspace/project

**Comment by cyrilletuzi** [comment]
Thanks. Can you point me to the Quokka issue/PR so I can check if it's the same issue?

**Comment by sukheja-varun** [comment]
Quokka provides it in their doc, have a look
https://quokkajs.com/docs/configuration.html#nodejs-version

**Comment by cyrilletuzi** [comment]
I need to see the code of the extension, so if you have the issue/PR link it would help a lot.

**Comment by sukheja-varun** [comment]
I am sorry but I don't have any git issue/PR
I can just tell you the quokka error message
```
Can not start node.js process (node), make sure your system has node.js installed.
```

**Comment by cyrilletuzi** [comment]
Well for now I would stay on the idea that the extension is not the issue. The error you reported in your first message has a lot of occurence on Google Search results, and it's always a problem of user config. Could you try some troubleshooting about your Node install (uninstall / reinstall, check it's your `PATH`, in your bash profile, etc.)?

**Comment by sukheja-varun** [comment]
With normal node installation, the path is fixed but with nvm, there are different versions of node installed on a system, and we keep changing the node version as per project need.
As quokka provides the solution for users using nvm instead of node directly as shown in https://quokkajs.com/docs/configuration.html#nodejs-version, you can too provide a param for providing node path for the plugin.

I can't set a particular node path for my whole system as then it will be no use of using nvm which keep changing the node version. VS code too provides the solution and it's terminal works well.

**Comment by cyrilletuzi** [comment]
There is no evidence at all that the issue is the same as in the Quokka extension (whose code is not publicy available).

Again, this extension is a VS Code extension, so the Node used is the one included in VS Code itself, as it's an Electron app, no matter your global installation. Then, it launches a command in the terminal, eitheir the default one, or the one specified in `terminal.integrated.shell.osx` or `terminal.integrated.shell.linux`. So your Node and npm must be properly configured in the according terminal. And then it just launches an Angular CLI command, trying to use the project local CLI, and if not available, the global CLI.

There is no steps in that where the extension could configure any Node path.

I'm not an expert of `nvm`, but while it allows to install multiple Node versions, it uses only one at a specific moment, with `nvm use`, and this step is supposed to change your `PATH` accordingly.

**Comment by cyrilletuzi** [comment]
There is nowhere in the lib where the `node` command is launched directly. It's the Angular CLI which must be do that, meaning that from where the CLI is launched, Node is not properly configured.

Problem is I can't reproduce your issue so it's not easy to help you. Can you tell me what happen when you do:
- `node -v`
- `npm -v`
- `ng version`
From:
- your zsh bash
- your default bash

This could help to check if it's the good terminal which is launched by the extension.

**Comment by sukheja-varun** [comment]
ZSH is the default bash and below are the result for commands:
- node -v => v12.2.0
- npm -v => 6.9.0
- ng version => 
 Angular CLI: 7.3.9
Node: 12.2.0
OS: linux x64
Angular: 
...

**Comment by cyrilletuzi** [comment]
By your default bash I mean the default terminal of your OS with the default bash (as one option could be that the extension open the wrong terminal).

Also, even if I don't think it's the issue, it could be worth a try to check with Node 10 instead of 12 (as Angular CLI doesn't officially support Node 12 yet).

**Comment by QBseventy** [comment]
Had the same issue, running node under control from nvm showed the same errors. 
"not-really" fixed it by installing corresponding nodejs (apt-get style) on ubuntu following instructions from https://github.com/nodesource/distributions/blob/master/README.md#deb

**Comment by cyrilletuzi** [comment]
@QBseventy Thanks for your help!

What do you mean by "not-really fixed"?

**Comment by sukheja-varun** [comment]
@cyrilletuzi Is it possible to change the default terminal for the plugin ?

@QBseventy nodejs (apt-get style) on ubuntu will download the node on ubuntu and in that case there will be no great use of node ?

**Comment by cyrilletuzi** [comment]
@sukheja-varun It is already: the extension uses the terminal configured in `terminal.integrated.shell.osx` or `terminal.integrated.shell.linux`.

**Comment by QBseventy** [comment]
> @QBseventy Thanks for your help!
> 
> What do you mean by "not-really fixed"?

Well as @sukheja-varun already mentioned the schematics now work for me, but nvm became a kind of useless.

**Comment by cyrilletuzi** [comment]
If it works when Node is installed the classic way, it's clearly a mismatch in configuration somewhere. I see 2 options:

- the only way where the extension could be responsible is if it doesn't launch the right terminal: that's quite unlikely because in the terminal detection done by the extension, Linux is the default. But I'll try to reproduce your environnement to be sure;

- otherwise it means there is a problem of configuration between nvm and the terminal you want, which is more likely due to the number of issues when googling your error, and in this case the extension can't do anything.

**Comment by cyrilletuzi** [comment]
Unfortunately I don't have a machine with Ubuntu at hand. I tried with Windows WSL but I can't debug extensions in VS Code with that. So I can't do the check and I'll need help from someone with Ubuntu.

Steps to reproduce in an Ubuntu with ZSH installed:
- in your OS terminal, go to the directory of your choice
- `git clone https://github.com/cyrilletuzi/vscode-angular-schematics/`
- open the created folder in VS Code
- set `terminal.integrated.shell.linux` to your `zsh` location (usually `/bin/zsh`) in your VS Code preferences
- restart VS Code to be sure
- open the VS Code terminal and check ZSH is used (prompt should start with `%`)
- add `console.log(userShell);` [here](https://github.com/cyrilletuzi/vscode-angular-schematics/blob/b345f583549ce74122699239f1f7cab358482055/src/schematics/utils.ts#L95)
- launch extension debug (4th icon on the left, below git icon, a bug striked ; then click the green play button)
- in the new opened window, open an Angular project
- in the same window, create a component with the extension
- go back in the primary window (with the extension project) and look at the message "debug" panel: should your zsh path (`/bin/zsh` or else); if it's `undefined`, the extension has an issue

**Comment by konieczny9lives** [comment]
Hi,

I am also using `zsh` and `nvm` on linux. I got the same error message as stated in this issue after upgrade to Angular 8 and uninstalling `nodejs` (debian package).

I suspected schematics package, but run above-mentioned steps by @cyrilletuzi and confirmed that the package is using the correct shell.

After reading https://github.com/Microsoft/vscode-docs/blob/master/docs/editor/integrated-terminal.md#why-is-nvm-complaining-about-a-prefix-option-when-the-integrated-terminal-is-launched I got to the root of the problem.

It turns out that when using `vscode` with `nvm` you have to remove global npm from `$PATH`. Then `vscode` on startup will use node from `nvm` correctly.

**Comment by cyrilletuzi** [comment]
@konieczny9lives Thank you for your help!

@sukheja-varun @QBseventy Is the solution proposed working for you?

**Comment by sukheja-varun** [comment]
```
/home/sukheja/.nvm/versions/node/v12.2.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```
This is my path and nowhere it contains default node, it just contains default nvm path

**Comment by t-zander** [comment]
I had the same issue. I am not sure what the reason is, but for me, solution was to launch vsCode from terminal, not from UI.
So I went to folder where my project was.
Entered
`code .`
If opened this way, the extension seems to work. Although I am not sure why it happens this way.

**Comment by sukheja-varun** [comment]
thanks @t-zander  it works for me as well
@cyrilletuzi  it works if vs-code is opened via terminal

**Comment by cyrilletuzi** [comment]
Thanks for the feedback!

Maybe it would be good to file an issue in VS Code repo.

**Comment by cyrilletuzi** [comment]
@sukheja-varun @QBseventy @konieczny9lives @t-zander Can you follow instructions in #56 and tell me if it solves this issue? The last VS Code update introduced a new API for extensions that may solve this problem.

**Comment by cyrilletuzi** [comment]
Oops, the issue number was wrong so there was no link, it's #55.

**Comment by cyrilletuzi** [comment]
Good news: this issue should be fixed by #125 (for real this time).

@sukheja-varun @QBseventy @konieczny9lives @t-zander  Can you follow the instructions in the PR to test and confirm the beta fixes the problem? Thanks.

**Comment by cyrilletuzi** [comment]
For info: v4 has just been released and is a full rewrite of the extension. Now the CLI command is launched in VS Code Terminal, so the extension should not be affected by custom configuration on user side, as long as `ng g component hello` works in your *VS Code* Terminal.

It took more than a *full-time* non-paid week to do this rewrite. Yet I still do not have any **[sponsor](https://github.com/sponsors/cyrilletuzi)**. So please consider to help, or to ask your enterprise, which makes money with this extension, to help.
